components/cdc/src/endpoint.rs

```diff
@@ -753,31 +823,156 @@ impl<T: 'static + RaftStoreRouter<RocksEngine>> Endpoint<T> {
+            let mut resp_incoming_voters = 0;
+            let mut resp_demoting_voters = 0;
+
+            region.get_peers().iter().for_each(|peer| {
+                let mut in_resp = false;
+                for store_id in &stores {
+                    if *store_id == peer.store_id {
+                        in_resp = true;
+                        break;
                     }
-                })
-                .collect();
-            match scheduler.schedule(Task::RegisterMinTsEvent) {
-                Ok(_) | Err(ScheduleError::Stopped(_)) => (),
-                // Must schedule `RegisterMinTsEvent` event otherwise resolved ts can not
-                // advance normally.
-                Err(err) => panic!("failed to regiester min ts event, error: {:?}", err),
+                }
+                match peer.get_role() {
+                    PeerRole::Voter => {
+                        voters += 1;
+                        if in_resp {
+                            resp_voters += 1;
+                        }
+                    }
+                    PeerRole::IncomingVoter => {
+                        incoming_voters += 1;
+                        if in_resp {
+                            resp_incoming_voters += 1;
+                        }
+                    }
+                    PeerRole::DemotingVoter => {
+                        demoting_voters += 1;
+                        if in_resp {
+                            resp_demoting_voters += 1;
+                        }
+                    }
+                    PeerRole::Learner => (),
+                }
+            });
+
+            let has_incoming_majority =
+                (resp_voters + resp_incoming_voters) >= ((voters + incoming_voters) / 2 + 1);
+            let has_demoting_majority =
+                (resp_voters + resp_demoting_voters) >= ((voters + demoting_voters) / 2 + 1);
+
+            has_incoming_majority && has_demoting_majority
+        };
+
+        // store_id -> leaders info, record the request to each stores
+        let mut store_map: HashMap<u64, Vec<LeaderInfo>> = HashMap::default();
+        // region_id -> region, cache the information of regions
+        let mut region_map: HashMap<u64, Region> = HashMap::default();
+        // region_id -> peers id, record the responses
+        let mut resp_map: HashMap<u64, Vec<u64>> = HashMap::default();
+        {
+            let meta = store_meta.lock().unwrap();
+            let store_id = meta.store_id.unwrap();
+            for (region_id, _) in regions {
+                if let Some(region) = meta.regions.get(&region_id) {
+                    if let Some((term, leader)) = meta.leaders.get(&region_id) {
+                        for peer in region.get_peers() {
+                            if peer.store_id == store_id {
+                                resp_map.entry(region_id).or_default().push(store_id);
+                                continue;
+                            }
```

Comment:
To forward resolved ts on this peer, we need to ensure this peer is the leader (maybe by checking `peer.store_id == store_id && peer.id == leader.id`) and the leader has applied on its term.